relayer/src/foreign_client.rs

```diff
@@ -406,6 +476,232 @@ impl ForeignClient {
+            // This can happen if a client update happened with a lower height than latest.
+            if target_height > update_event.consensus_height() {
+                continue;
+            }
+
+            // Ensure consensus height of the event is same as target height. This should be the
+            // case as we either
+            // - got the `update_event` from the `target_height` above, or
+            // - an `update_event` was specified and we should eventually find a consensus state
+            //   at that height
+            // We break here in case we got a bogus event.
+            if target_height < update_event.consensus_height() {
+                break;
+            }
+
+            // Check for misbehaviour according to the specific source chain type.
+            // In case of Tendermint client, this will also check the BFT time violation if
+            // a header for the event height is not available.
+            let misbehavior = self
+                .src_chain
+                .build_misbehaviour(client_state.clone(), update_event.clone())
+                .map_err(|e| {
+                    ForeignClientError::Misbehaviour(format!("failed to build misbehaviour {}", e))
+                })?;
+
+            if misbehavior.is_some() {
+                // TODO - add updateClient messages if supporting headers are missing on-chain
+                return Ok(misbehavior);
+            }
+
+            // Clear the update
+            update = None;
+        }
+
+        Ok(None)
+    }
+
+    pub fn detect_misbehaviour_and_send_evidence(
+        &self,
+        update: Option<UpdateClient>,
+    ) -> Result<Option<IbcEvent>, ForeignClientError> {
+        match self.handle_misbehaviour(update)? {
+            None => Ok(None),
+            Some(misbehaviour) => {
+                error!("MISBEHAVIOUR detected {}, sending evidence", misbehaviour);
+
+                let signer = self.dst_chain().get_signer().map_err(|e| {
+                    ForeignClientError::Misbehaviour(format!(
+                        "failed getting signer for dst chain ({}) with error: {}",
+                        self.dst_chain.id(),
+                        e
+                    ))
+                })?;
+
+                let msg = MsgSubmitAnyMisbehaviour {
+                    client_id: self.id.clone(),
+                    misbehaviour,
+                    signer,
+                };
+
+                let events = self
+                    .dst_chain()
+                    .send_msgs(vec![msg.to_any()])
+                    .map_err(|e| {
+                        ForeignClientError::Misbehaviour(format!(
+                            "failed sending evidence to dst chain ({}) with err: {}",
+                            self.dst_chain.id(),
+                            e
+                        ))
+                    })?;
+
+                Ok(Some(events[0].clone()))
```

Comment:
Are we guaranteed to have at least one event here? If not, we should probably add a check otherwise this code will panic. (I hate how Rust usually makes potential panics very explicit (`unwrap` etc.) _except_ for indexing into vectors…)

Reply:
Good point! We should get some event the way we have the code now. But in the future we might prepend some client updates before the misbehaviour message. I changed the code to return all events.